_posts/2019-07-25-testy-frontendu-okiem-full-stacka.md

```diff
@@ -0,0 +1,136 @@
+    - angular
+---
+
+Niełatwo znaleźć wymówkę, żeby nie pisać testów jednostkowych. Obecność frameworków ułatwiających tę czynność w projektach, z którymi stykamy się na co dzień, nie powinna na żadnym chociaż trochę doświadczonym programiście robić wrażenia. Nie trzeba pewnie Cię przekonywać, że **jedne z wielu zalet pisania testów jednostkowych**, to:
+* zmuszenie twórcy do zastanowienia się nad zadaniem sprawdzanego kodu (co może potencjalnie poprawić architekturę aplikacji),
+* ułatwienie wczesnego wyłapywania błędów,
+* ekspozycja przypadków brzegowych,
+* ułatwienie zrozumienia działania kodu osobom, które go nie tworzyły.
+
+W wielu nowoczesnych aplikacjach internetowych, w tym np. we wnioskach Eximee, **duża część logiki znajduje się po stronie klienckiej**, hipokryzją byłoby pominięcie testów w tak istotnym elemencie aplikacji, bo przecież wiadomo, że najszybciej wychodzą na jaw błędy właśnie tam. Niemniej nawet przy ogromnej liczbie narzędzi wspomagających proces pisania testów jednostkowych, programiści mogą mieć problem z **wyznaczeniem właśnie tych jednostek**.
+## Co testować
+Niezależnie od tego jaki framework został użyty w danym projekcie, zawsze możemy z niego wydzielić komponenty. Przeważnie jest to JavaScriptowa klasa z jakimś odniesieniem do szablonu HTML. W praktyce możemy podzielić te komponenty na dwa typy: 
+* **komponent prezentacyjny**, który **nie posiada logiki biznesowej** i jego jedynymi zadaniami są wyświetlenie szablonu na podstawie wejścia i ew. przekazanie jakiegoś zdarzenia (np. kliknięcia, wciśnięcia klawisza itp.) do komponentu nadrzędnego,
+* komponent, który **używa i zarządza innymi komponentami** nie zajmując się jednocześnie zbytnio prezentacją.
+
+Brak tego podziału może znacząco utrudnić pisanie testów jednostkowych, co zresztą okaże się bardzo szybko przy próbie napisania ich do słabo zaprojektowanego komponentu.
+Uważam, że testowanie komponentów prezentacyjnych jest zasadne tylko w przypadku, gdy wejście w jakiś sposób zmienia jego zachowanie lub obsługa uaktualnienia widoku jest skomplikowana (np. animacja przeliczając atrybuty elementu w locie). **Najważniejsze jest zwiększenie pokrycia logiki biznesowej.**
+
+## Testowanie logiki biznesowej
+W pierwszej kolejności powinniśmy się zastanowić nad tym, czy z komponentu możemy wydzielić logikę, np. do osobnego serwisu lub serwisów. Serwisy testuje się o wiele prościej niż komponenty, ze względu na brak szablonu i jasno zdefiniowane wejścia i wyjścia metod. Niech przykładem będzie komponent wybory daty z formatterem - zakładając, że cała logika znajduje się w komponencie, trzeba będzie zadbać o stworzenie jego instancji ze wszystkimi zależnościami pisząc testy dla formattera, następnie zasymulować zdarzenie wpisania danych w pole tekstowe. Gdyby wydzielono wcześniej osobny serwis do formatowania, to wystarczyłoby przetestować tylko jego logikę. Testy całego komponentu możemy przeprowadzić zaślepiając odpowiednie zależności, co znacznie ułatwi pracę.
```

Comment:
> W pierwszej kolejności powinniśmy się zastanowić nad tym, czy z komponentu możemy wydzielić logikę, np. do osobnego serwisu lub serwisów. Serwisy testuje się o wiele prościej niż komponenty, ze względu na brak szablonu i jasno zdefiniowane wejścia i wyjścia metod.

1. Warto by zdefiniować co rozumiesz przez "serwis" (tak jak zrobiłeś to z "komponentami") - zakładałbym, że odbiorcą artykułu może być niedoświadczony frontendowiec, który nigdy nie słyszał takiego pojęcia
2. Czy "komponent" nie ma jasno zdefiniowanych wyjść i wejść?

Reply:
1. rozwinę o co chodzi
2. no tu popłynąłem trochę, zmienię to